src/net.cpp

```diff
@@ -3779,6 +3832,24 @@ uint64_t CConnman::CalculateKeyedNetGroup(const CAddress& address) const
     return GetDeterministicRandomizer(RANDOMIZER_ID_NETGROUP).Write(vchNetGroup).Finalize();
 }
 
+void CConnman::PerformReconnections()
+{
+    AssertLockNotHeld(m_reconnections_mutex);
+    AssertLockNotHeld(m_unused_i2p_sessions_mutex);
+    while (true) {
+        // Move first element of m_reconnections to todo (avoiding an allocation inside the lock).
+        decltype(m_reconnections) todo;
+        {
+            LOCK(m_reconnections_mutex);
+            if (m_reconnections.empty()) break;
+            todo.splice(todo.end(), m_reconnections, m_reconnections.begin());
+        }
+
+        auto& [addr_connect, count_failure, grant, dest, conn_type] = *todo.begin();
+        OpenNetworkConnection(addr_connect, count_failure, &grant, dest.empty() ? nullptr : dest.c_str(), conn_type);
```

Comment:
I think this means that reconnections for added nodes can be made from `ThreadOpenConnections`, and reconnections for automatic connections can be performed by `ThreadOpenAddedConnections` - which seems really weird, although I don't think it breaks anything. I wonder if it makes sense to perform the respective reconnection attempts by their designated threads.

Reply:
Ok - as pointed out below it could also help resolve problems with a blocking semaphore to have another thread trying as well.